packages/website/blog/2024-08-15-parser-options-project-service.mdx

```diff
@@ -0,0 +1,260 @@
+
+- `allowDefaultProjectForFiles`: a glob of "out-of-project" files to lint with type information
+- `defaultProject`: path to a TSConfig to use for out-of-project file type information
+
+For example, the following config solves the common case of projects that have root-level files like `eslint.config.js` and `vitest.config.ts`:
+
+<Tabs groupId="eslint-config">
+<TabItem value="Flat Config">
+
+```js title="eslint.config.js"
+export default tseslint.config({
+  // ...
+  languageOptions: {
+    parserOptions: {
+      projectService: {
+        allowDefaultProjectForFiles: ['./*'],
+        defaultProject: './tsconfig.json',
+      },
+      tsconfigRootDir: import.meta.dirname,
+    },
+  },
+  // ...
+});
+```
+
+</TabItem>
+<TabItem value="Legacy Config">
+
+```js title=".eslintrc.cjs"
+module.exports = {
+  // ...
+  parser: '@typescript-eslint/parser',
+  parserOptions: {
+    projectService: {
+      allowDefaultProjectForFiles: ['./*'],
+      defaultProject: './tsconfig.json',
+    },
+    tsconfigRootDir: __dirname,
+  },
+  // ...
+};
+```
+
+</TabItem>
+</Tabs>
+
+:::tip
+This means you should be able to remove any lint-only `tsconfig.eslint.json` files!
+🥳
+:::
+
+## 🧠 Predictability and Reliability
+
+One of the challenges of typed linting is using TypeScript APIs designed for TypeScript projects in ESLint's runtime.
+[ESLint does not provide parsers session information](https://github.com/eslint/rfcs/pull/102).
+As a result, `@typescript-eslint/parser` has to use approximations to guess whether it was in the more performance-friendly "single-run" mode (rather than in ESLint's multi-pass `--fix` mode or a long-lived editor session).
+Single-run mode comes with several bugs.
+For example:
+
+- ESLint's `--fix` mode breaks type information after the first run ([#9577](https://github.com/typescript-eslint/typescript-eslint/pull/9577))
+- Extra file extensions, such as those used by `.svelte` and `.vue` fileSync, are not supported ([#9504](https://github.com/typescript-eslint/typescript-eslint/issues/9504))
+
+typescript-eslint's single-run inference enables uses common heuristics such as checking for `'--fix'` in `process.argv`, the presence of `process.env.CI`, and the presence of `parserOptions.extraFileExtensions`.
+It can be disabled with `parserOptions.disallowSingleRunInference`.
+
+Enabling single-run mode generally improves performance by 10-20%.
+typescript-eslint@v8 enables inference of single-run mode by default.
+If your project is stuck on `parserOptions.project`, we recommend keeping single-run inference on if possible.
+
+`parserOptions.projectService` does not suffer from the same bugs as `parserOptions.project` with single-run mode.
+It supports extra file extensions out-of-the-box and does not slow down when used with ESLint's `--fix`.
+
+We recommend switching to `parserOptions.projectService` if possible.
+
+## ⚡️ Performance Comparisons
```

Comment:
So I don't forget: https://github.com/t3-oss/create-t3-app/pull/1936#discussion_r1667389041 is an example of a perf improvement in the wild.